File: web/src/types/crypto.ts
export interface CryptoSnapshot {
  symbol: 'BTCUSDT'; evaluatedAt: string; regime: 'BULL' | 'BEAR' | 'RANGE' | 'UNKNOWN';
  setup: 'BREAKOUT' | 'NONE'; action: 'BUY' | 'WAIT' | 'HOLD' | 'EXIT'; price: string;
  // Digit/letter boundaries follow the shared camelcase-keys API adapter.
  ema201H: string | null; ema501H: string | null; ema2015M: string | null;
  breakoutLevel15M: string | null; volumeRatio15M: string | null; atr1415M: string | null;
  initialStop: string | null; trailingStop: string | null;
  positionBefore: string | null; positionAfter: string | null; positionDelta: string | null;
  averageEntryPrice: string | null;
  positionState: 'FLAT' | 'LONG'; reason: string;
}
export interface CryptoState {
  symbol: 'BTCUSDT'; positionState: 'FLAT' | 'LONG'; entryPrice: string | null;
  positionPct: string; averageEntryPrice: string | null;
  entryTime: string | null; highestPriceSinceEntry: string | null;
  initialStop: string | null; trailingStop: string | null; updatedAt: string | null;
}
export interface CryptoOverview {
  symbol: 'BTCUSDT'; strategy: CryptoSnapshot | null; state: CryptoState;
}

export interface CryptoExecution {
  evaluatedAt: string; price: string; positionBefore: string; positionAfter: string;
  positionDelta: string; action: string; reason: string;
}
export interface CryptoTrade {
  entryTime: string; exitTime: string; holdingSeconds: number;
  averageEntryPrice: string; exitPrice: string; realizedReturn: string;
}
export interface CryptoPerformance {
  performanceStartAt: string | null; performanceEndAt: string | null;
  currentPosition: { positionPct: string; averageEntryPrice: string | null };
  executionCount: number; completedCycles: number; winCount: number; lossCount: number;
  winRate: string | null; averageReturn: string | null; cumulativeReturn: string; maxDrawdown: string;
  bestTrade: string | null; worstTrade: string | null;
  recentExecutions: CryptoExecution[]; recentTrades: CryptoTrade[];
  equityCurve: { evaluatedAt: string; equity: string; drawdown: string }[];
}

File: src/finance_analysis/interfaces/api/v1/schemas/crypto.py
"""BTC strategy REST contracts. Decimal JSON values are strings."""

from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel


class CryptoStateResponse(BaseModel):
    symbol: Literal["BTCUSDT"] = "BTCUSDT"
    position_state: Literal["FLAT", "LONG"]
    entry_price: Decimal | None = None
    entry_time: datetime | None = None
    highest_price_since_entry: Decimal | None = None
    initial_stop: Decimal | None = None
    trailing_stop: Decimal | None = None
    updated_at: datetime | None = None


class CryptoSnapshotResponse(BaseModel):
    symbol: Literal["BTCUSDT"] = "BTCUSDT"
    evaluated_at: datetime
    regime: Literal["BULL", "BEAR", "RANGE", "UNKNOWN"]
    setup: Literal["BREAKOUT", "NONE"]
    action: Literal["BUY", "WAIT", "HOLD", "EXIT"]
    price: Decimal
    ema20_1h: Decimal | None = None
    ema50_1h: Decimal | None = None
    ema20_15m: Decimal | None = None
    breakout_level_15m: Decimal | None = None
    volume_ratio_15m: Decimal | None = None
    atr14_15m: Decimal | None = None
    initial_stop: Decimal | None = None
    trailing_stop: Decimal | None = None
    position_state: Literal["FLAT", "LONG"]
    reason: str


class CryptoOverviewResponse(BaseModel):
    symbol: Literal["BTCUSDT"] = "BTCUSDT"
    strategy: CryptoSnapshotResponse | None
    state: CryptoStateResponse


class CryptoSignalsResponse(BaseModel):
    items: list[CryptoSnapshotResponse]
